### PR TITLE
Add /v1/users/login 

### LINKED ISSUE #5: U2: Visitor can login using email and password.
```javascript
Endpoint:
POST /v1/users/login
{
  email: String,
  password: String,
}

Response:
Success:
{status: 'SUCCESS', payload: '<Token>'}
Error:
{status: 'ERROR', payload: '<Error Message>'}
```

### PR DESCRIPTION
**Pull request by HinataIzayoi**
Process POST request from /v1/users/login.

Return API error if input email does not exist or have wrong email password combination.
Create a Token and return API success and Token if succeed.

Resolve #5

### CODE REVIEW

File: controllers/userController.js [hunk 1]
```diff
@@ -1,6 +1,8 @@
 const User = require('../models/User');
 const {apiError, apiSuccess, genPasswordSalt, calcPasswordHash} = require('./utils');
 
+const tokens = {};
```
Comment (Kailang) [inline controllers/userController.js]: Create a new Node module called "tokenService" under /services/tokenService.js that supports the following operation:
1. createToken(userId: String): void
2. checkTokenValid(token: String): Boolean
3. getUserId(token: String): String|undefined

File: controllers/userController.js [hunk 2]
```diff
@@ -23,3 +25,20 @@ exports.register = async function(params) {
...
+
+
+exports.login = async function(params) {
+  const user = await User.findOne({email: params.email});
+  if (!user) {
+    return apiError('Invalid email');
+  }
+  const hash = calcPasswordHash(params.password, String(user.passwordSalt));
+  if (hash === user.passwordSha256 ) {
+    const token = genPasswordSalt();
```
Comment (Kailang) [inline controllers/userController.js]: Rename genPasswordSalt() to genSecureRandomString(length: Number|undefined).
Make genSecureRandomString() return the random string with default length.
And, make genSecureRandomString(length) return the random string with length `length`

File: controllers/userController.js [hunk 3]
```diff
@@ -23,3 +25,20 @@ exports.register = async function(params) {
...
+    tokens[token] = user.userName;
+    console.log(tokens);
```
Comment (Kailang) [inline controllers/userController.js]: Remove console.log().

File: routes/v1/usersRouter.js [hunk 4]
```diff
@@ -1,6 +1,7 @@
 const express = require('express');
 const router = express.Router();
 const userController = require('../../controllers/userController');
+const User = require('../../models/User');
```
Comment (Kailang) [inline routes/v1/usersRouter.js]: Why do you need to access the Model in router?

File: controllers/userController.js [hunk 5]
```diff
@@ -23,3 +25,20 @@ exports.register = async function(params) {
...
   return apiSuccess();
 };
+
+
+exports.login = async function(params) {
+  const user = await User.findOne({email: params.email});
+  if (!user) {
+    return apiError('Invalid email');
+  }
+  const hash = calcPasswordHash(params.password, String(user.passwordSalt));
```
Comment (Kailang) [inline controllers/userController.js]: String(user.passwordSalt) is not necessary.
Just user.passwordSalt should be sufficient.

File: controllers/userController.js [hunk 6]
```diff
@@ -23,3 +25,20 @@ exports.register = async function(params) {
...
+  if (hash === user.passwordSha256 ) {
+    const token = genPasswordSalt();
+    tokens[token] = user.userName;
+    console.log(tokens);
+    return apiSuccess(token);
+  }
+
+  return apiError('Invalid combination of email and password');
```
Comment (Kailang) [inline controllers/userController.js]: Move the error branch before the success branch.
We handle the error branches first.

File: controllers/userController.js (hunk 1 above)
Comment (HinataIzayoi) [reply controllers/userController.js]: If createToken is void, how to return token in response? Would it make more sense to return generated token instead?

File: controllers/userController.js (hunk 5 above)
Comment (HinataIzayoi) [reply controllers/userController.js]: Acknowledged.

File: controllers/userController.js (hunk 6 above)
Comment (HinataIzayoi) [reply controllers/userController.js]: Acknowledged.

File: controllers/userController.js (hunk 2 above)
Comment (HinataIzayoi) [reply controllers/userController.js]: Done.

File: controllers/userController.js (hunk 3 above)
Comment (HinataIzayoi) [reply controllers/userController.js]: Done.

File: routes/v1/usersRouter.js (hunk 4 above)
Comment (HinataIzayoi) [reply routes/v1/usersRouter.js]: Used for debugging purposes. Have removed.

**Pull request closed without merging**